### DISCUSSION TITLE
TS tests

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#8392] [General]
## Motivation

#8297 types are complex. That introduced the need to test typing, see TS playgrounds in comments.
For example I didn't expect the following to fail:
The `onChange` callback is not inferred correctly when using an array

![image](https://user-images.githubusercontent.com/34343793/197605448-afc9fc8b-91b1-44c7-8700-135bc6090a82.png)

I worked on types. And then I had to test it comes out as I expected it.
So I found myself writing calls to the animate function to see if I wrote the complex type correctly. 
Instead of writing and deleting why not have a dedicated place that I can work on, comment and discuss in a PR and make sure it behaves as I expect it to.
Then, regardless, I found that the entry point to the repo is not typed (I remember seeing an issue about it). A good case for why ts tests are crucial.

![image](https://user-images.githubusercontent.com/34343793/197469717-211ec232-b747-41d3-b177-2e2d5781958e.png)

## Research
- https://stackoverflow.com/questions/49296151/how-to-write-tests-for-typescript-typing-definition:
  - suggests `dtslint`, `tsd`, see below
  - suggests using `// @ts-expect-error` and test if the file compiles.
  - refers to how [express tests TS](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express/express-tests.ts)
- https://github.com/Microsoft/dtslint => `@definitelytyped/dtslint`
suggested by official TS but has limitation since it does a string comparison of the compiler returned type so `a | b != b | a` and typos in test files will f*** us over
Seems to run only on declaration files so is completely worthless
- https://github.com/SamVerschueren/tsd
js test suite with `expectType`, `expectError` etc., suggested by official TS
Can only assert a value, not params. so is worthless on its own
- https://github.com/mmkal/expect-type
js test suite, seems not popular yet
- https://github.com/dsherret/conditional-type-checks
type assertions + js assertions, looks good
- https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type
unrelated to testing

## Findings

- use `// @ts-expect-error explanation` to assert an error and test for file compilation
- `conditional-type-checks` | `tsd` => `conditional-type-checks` wins since it is more flexible and accurate (see playground below) and allows testing parameters and whatever
- `@definitelytyped/dtslint` is useless

**Comment by ShaMan123** [comment]
Another interesting repo
https://github.com/woutervh-/typescript-is

Seems it can help in the source code as well, see `assertType`

**Comment by ShaMan123** [comment]
```ts
import { IsExact, assert } from 'conditional-type-checks';

type Base<T extends number | number[]> = |
    ({
        endValue?: T,
        byValue?: never,
    }|
    {
        endValue?: never,
        byValue?: T,
    }) &
    {
        onChange?: (a: T, b: number, c: number) => void
    }


function animate<T extends Base<number> | Base<number[]>>(options: T) {
    return {
        endValue: null as unknown as T
    }
}

animate({
    endValue: 3,
});
animate({
    byValue: 2,
});
// @ts-expect-error only one of (`byValue` | `endValue`) is allowed
animate({
    endValue: 3,
    byValue: 2,
});

function assertStrict<T, U>(assertTrue: IsExact<T, U>) {
    
}

const context = animate({
    startValue: 1,
    endValue: 3,
    onChange(
        a, 
        b,
        c
    ) { 
        assertStrict<typeof a, number>(true);
        assertStrict<typeof b, number>(true);
        assertStrict<typeof c, number>(true);
    }
})

const arrayContext = animate({
    startValue: [5],
    byValue: [1],
    onChange(
        a,
        b,
        c
    ) {
        assertStrict<typeof a, number[]>(true);
        assertStrict<typeof b, number[]>(true);
        assertStrict<typeof c, number[]>(true);
    }
});

```



<details><summary><b>Output</b></summary>

```ts
function animate(options) {
    return {
        endValue: null
    };
}
animate({
    endValue: 3,
});
animate({
    byValue: 2,
});
// @ts-expect-error only one of (`byValue` | `endValue`) is allowed
animate({
    endValue: 3,
    byValue: 2,
});
function assertStrict(assertTrue) {
}
const context = animate({
    startValue: 1,
    endValue: 3,
    onChange(a, b, c) {
        assertStrict(true);
        assertStrict(true);
        assertStrict(true);
    }
});
const arrayContext = animate({
    startValue: [5],
    byValue: [1],
    onChange(a, b, c) {
        assertStrict(true);
        assertStrict(true);
        assertStrict(true);
    }
});
export {};

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ssl=1&ssc=19&pln=1&pc=25#code/JYWwDg9gTgLgBAbzgSQM4FEAeBDAxjAGjm1VQFNY4BfOAMyghDgHJcIA7AE2BmA+wA2AWhgBPMGSG4AFmVwBrVMwDcAWABQGsRLgAhEmQA8AFThlMMMl1Rx2AVxAAjCnAA+th86gBtALoA+OABeNw04cLgACgQwiLirTgA1QTsyAH4ALjhjAli48MdRZIFUzNsyADcKXPU4qlc8xEb4rmLSrPZK6uaIwrb0rJzGqgBKOAAyRpja-PCOAGFpbHYAcwGo7EGiRw7ParhcXacKMaDAiohgTmGNW-VaO3Z8PnZidlBsSxMzCwSbfXIhnsxyggXcAKMwK8fn8-kiEDAvA4qEGY2mcSgZBgdigr3RszMrRSZF2AgExBsj3k7AgAHdXiRsjd1FQ7ssPpZoo0Ev0sgBmGqjNTqdkgT5kLkzApFYlZABMgpGwoA9Mq4AABGCoITmCT4HVQBhQOAcASiE2dE20KIAAz6xJtbjgNp5DrGwBsggEdLI1xF7zFnPx4VdJRJcAFjXtYfliuFGgeTyRDNIFBgAGUYFBgPgTEQAKpwkjkWDGKCpLJoLB4GB5uCFtGNDSszTqNjsVDwduWCzBN4ciXBuCd7CwXlwACMNRaSVlEenEQWS1WEp64WwRDXcEcC9muEaaLgW+Lacz2dz2jIEGtG48ILhWdSSuPqdgZ5ztcv1+3RChFAf5ZkM+Ur5Ceb5Zh+hhftauC-nsoKRI+QHCnUzYjHc7adsQhrYKI8wcD28AhKK4qSnEI5jnO3gAKy+Lu24yjGcDeBOdGNEuyxrJEx70XEO5bvuIGNiBcRgRmEEXuIV43nBIIwohgHAQSFIluJ56flJ347ne0IBApT4oQSYnvpJEjfrBOkUPJSFKRELZChoQA)

**Comment by ShaMan123** [comment]
```ts
import { expectType } from 'tsd';

type Base<T extends number | number[]> = |
    ({
        endValue?: T,
        byValue?: never,
    }|
    {
        endValue?: never,
        byValue?: T,
    }) &
    {
        onChange?: (a: T, b: number, c: number) => void
    }


function animate<T extends Base<number> | Base<number[]>>(options: T) {
    return {
        endValue: null as unknown as T
    }
}

animate({
    endValue: 3,
});
animate({
    byValue: 2,
});
// @ts-expect-error only one of (`byValue` | `endValue`) is allowed
animate({
    endValue: 3,
    byValue: 2,
});

const context = animate({
    startValue: 1,
    endValue: 3,
    onChange(
        a, 
        b,
        c
    ) { 
        expectType<number>(a)
        expectType<number>(b)
        expectType<number>(c)
    }
})

const arrayContext = animate({
    startValue: [5],
    byValue: [1],
    onChange(
        a,
        b,
        c
    ) {
        // should be marked as an error
        expectType<number[]>(a)
        expectType<number[]>(b)
        expectType<number[]>(c)
    }
});

```



<details><summary><b>Output</b></summary>

```ts
import { expectType } from 'tsd';
function animate(options) {
    return {
        endValue: null
    };
}
animate({
    endValue: 3,
});
animate({
    byValue: 2,
});
// @ts-expect-error only one of (`byValue` | `endValue`) is allowed
animate({
    endValue: 3,
    byValue: 2,
});
const context = animate({
    startValue: 1,
    endValue: 3,
    onChange(a, b, c) {
        expectType(a);
        expectType(b);
        expectType(c);
    }
});
const arrayContext = animate({
    startValue: [5],
    byValue: [1],
    onChange(a, b, c) {
        // should be marked as an error
        expectType(a);
        expectType(b);
        expectType(c);
    }
});

```


</details>


<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```


</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUwB5mQYxgFQJ4ZwC+cAZlBCHAOQwDOAJtQNwCwAUBzAcnAEIBDOsgA8OFKhjIAdgzpxpAVxAAjZFDgAfBcrVQA2gF0AfHAC8WjnGtwAFAis2nMhgDUBAG0XIA-AC44HAAaRydrFTx3L18A6WQAN3UQ9iciTVDEDOdZKO9-BQSkrJsI3JjA5NSASjgAMgyHFLDrCGkAYQALAWkAc3LbAQDguBVY3SS4TDHVdRqzU3iIYAYMog519lJFaWxgVrhu0AEpMQkpWXlBYRElGahTbSvRW70jY2NbCDAYPek6IZqjScUGQMEUUGkmSazRcZTGHg8B3k2wA1tIIAB3SFCQKrDhrTjsQ4gY7IewZWGebwBADMySIVTYROkRyk5OhI0iVOQAQATPTGRwAPRCuAAAXoAFo0BhsNKoBQNK0PHg4K1eBBSHYAAalbnarRwbWU6LamrAeSeDyY5ArZmsslAmwm6lwOkZPXRPkCpkcTCtOjwf3SKSScwHFkktlO6yBgSwOFwACMlWdOW5tNTLXaXV6ZOK1gEQTgBZGWbCmAygJLHOc6CwuB4N3G9wGVVLMob+AwzbuHxU7drzvr2G7zxbH0wg9S+MHfoD8HjUAEeDarVD8AsxNJ7KccYTGbg+gArIZy57Xfok2eMq1Ot0+rZS0XSypy05KxzAaWRXA6B0IEUDwGBGXgSSgFFbSRCMUAVaAOxHRsexedQ3jbBDZSQ8c7jQgcMK7JsUIMExbCnPF2AZJkgA)

**Comment by ShaMan123** [comment]
@Lazauya want to join?

**Comment by ShaMan123** [comment]
updated description regarding `dtslint` which is completely useless

**Comment by ShaMan123** [comment]
PR #8394

**Comment by asturur** [comment]
@ShaMan123 i expected something else.
Can you make a clear example of what we expect to test and why, and also why a simple ts -noEmit is not good enough for verifying types have been handled correctly?

I would expect that Rollup being able to transpile the code correctly is good enough.

**Reply by ShaMan123** [reply]
Updated description

**Comment by jiayihu** [comment]
Just write your tests in TS and the typings will be tested automatically as well, no need for type-only tests in this case since fabric is not a type-level library. You also get two birds with one stone: testing the types and testing the implementation with the same test.

**Reply by gloriousjob** [reply]
Assuming jest is the test framework, they say we should use ts-jest instead of jest for some typescript support: https://jestjs.io/docs/getting-started

Alternatively, could we consider a similar but typescript supporting framework like vitest?

**Reply by ShaMan123** [reply]
Vitest was considered. I recall @asturur dismissing it as too immature.
Can you list why it is better?

**Reply by gloriousjob** [reply]
Main thing is typescript support out of the box. I think it was basically made because jest took forever to support vite so the other pro is greater likelihood that you'll get a response to an issue but it's almost a drop in replacement for jest as far as test structure.
Agreed on maturity. They just started putting out v1 betas so it’s starting to mature.

**Reply by ShaMan123** [reply]
In general I dislike facebook (in general) open source (jest) after being a react-native dev
I distrust their commitment to open source

**Reply by asturur** [reply]
my point on jest is just that is something that everyone knows.
Why do we think jest has no ts support?
It uses babel and is painfully slow, yes.
What is jest missing for typescript?

**Reply by gloriousjob** [reply]
I think this is the feature that the discussion is focusing on:
https://vitest.dev/guide/testing-types.html
